pkg/controller/jobs/pod/pod_controller.go

```diff
@@ -739,7 +739,8 @@ func (p *Pod) notRunnableNorSucceededPods() []corev1.Pod {
 // isPodRunnableOrSucceeded returns whether the Pod can eventually run, is Running or Succeeded.
 // A Pod cannot run if it's gated and has a deletionTimestamp.
 func isPodRunnableOrSucceeded(p *corev1.Pod) bool {
-	if p.DeletionTimestamp != nil && len(p.Spec.SchedulingGates) > 0 {
+	isPendingWithoutNode := p.Status.Phase == corev1.PodPending && p.Spec.NodeName == ""
+	if p.DeletionTimestamp != nil && (isPendingWithoutNode || len(p.Spec.SchedulingGates) > 0) {
```

Comment:
Why not just `if p.DeletionTimestamp != nil && p.Spec.NodeName == ""`? 

We know there are no `schedulingGates` if `p.Spec.NodeName != ""` because of this [check](https://github.com/kubernetes/kubernetes/blob/5ceb99dc6bae8a3013192439bdbee942d14de190/pkg/apis/core/validation/validation.go#L4987). Also, a pod with `p.Spec.NodeName == ""` is Pending (or Failed), so no need to check for Pending explicitly.

This would also match the logic in `gc_controller` [link](https://github.com/kubernetes/kubernetes/blob/5ceb99dc6bae8a3013192439bdbee942d14de190/pkg/controller/podgc/gc_controller.go#L303).

For the context, the users may want to implement the analogous logic in their controllers which create replacement pods, so that the replacement pods are safe (not handled as excess pods). So, the simpler the logic, the easier it would be for users to replicate.